pkg/bspatch/byte_add_reader.go

```diff
@@ -0,0 +1,59 @@
+package bspatch
+
+import (
+	"errors"
+	"fmt"
+	"io"
+)
+
+var ErrUnequalReads = errors.New("byteAdder.Read: did not read equal number of bytes from both readers")
+
+type byteAddReader struct {
+	r1   io.Reader
+	r2   io.Reader
+	addb []byte
+}
+
+func newByteAddReader(r1 io.Reader, r2 io.Reader) byteAddReader {
+	return byteAddReader{
+		r1:   r1,
+		r2:   r2,
+		addb: make([]byte, ReadBufferSize),
+	}
+}
+
+// Read adds the bytes of the two readers and writes to using p as scratch space.
+func (ba byteAddReader) Read(p []byte) (written int, err error) {
+	chunk := min(len(p), ReadBufferSize)
+
+	for written < len(p) {
```

Comment:
I think it's much simpler to use only `p`: the first half can be used to read from one reader, and the second half from another.

This means the caller may have to make twice as many calls to `Read`, but then you don't have to do your own buffering, and can just rely on the buffering by the client (in this case `io.Copy`).

Also, it becomes much simpler:

```
limit := len(p)/2
n1, err1 := ba.r1.Read(p[:limit])
n2, err2 := io.ReadFull(r2, p[limit:limit + n1])
for i, b := range p[limit:limit + n2] {
    p[i] += b
}

// err := your error handling
return n2, err
```

Reply:
This was very thought provoking. The suggested change is a net complexity and memory win, and there seems to be no loss in error transparency since reads would be looped in either case. But it's a less natural abstraction to me at the moment. 

I would imagine the general case of N readers would use an internal buffer instead of fragmenting p into `len(p)/N` and shifting the responsibility to the caller to make N reads+writes. We're not really dealing with that case, and readers don't need to "best effort" read `len(p)` bytes as far as `io.Reader` specifies. Composability seems to indicate that client should be the one to specify buffering behavior e.g. composing this reader with a `bufio.Reader` or `io.Copy` with a `bufio.Writer` to reduce flushing / small reads instead of anticipating as this PR currently stands. Would you say that's accurate? 

Would appreciate any thoughts around how to think about this / resources for what might be idiomatic.

Reply:
I think the part I'm having a hard time wrapping my head around is forgoing an internal best effort at filling `p`.

Reply:
(This also imply that `len(p) === 1` would be read and handled conditionally, which is less elegant but simple to reason about)

Reply:
Yeah, I struggled with this for a bit too. I think fundamentally there is no way to make the `io.Reader` abstraction satisfy all the properties we would like:

1. buffer sizes are handled by the caller
2. we fill as much of `p` as we can efficiently do without doing I/O in a loop

My approach forgoes (2), while your approach forgoes (1). I think forgoing (2) is better because:
1. it's overall slightly more memory efficient
2. it's more CPU-efficient due to the buffer sizes for each reader matching exactly
3. it is less complicated to implement
4. no contract is written down anywhere about how much of `p` we need to fill.

I'm not sure (1) is a particularly big win, but (2) is substantial.

Side note: my implementation breaks if the passed in buffer has size 1, since then we won't read anything, and we definitely should eventually read stuff. Luckily, that won't happen due to the way we use it, but maybe worth leaving a comment.